src/scheduler/Scheduler.cpp

```diff
@@ -95,16 +95,29 @@ void Scheduler::reset()
     resetThreadLocalCache();
 
     // Shut down all Executors
-    for (auto& p : executors) {
+    // Executor shutdown takes a lock itself, so "finish" executors without the
+    // lock.
+    decltype(executors) executorsList;
+    decltype(deadExecutors) deadExecutorsList;
+    {
+        faabric::util::FullLock lock(mx);
+        executorsList = executors;
+    }
+    for (auto& p : executorsList) {
         for (auto& e : p.second) {
             e->finish();
         }
     }
-
+    executorsList.clear();
+    {
+        faabric::util::FullLock lock(mx);
+        deadExecutorsList = deadExecutors;
+    }
     for (auto& e : deadExecutors) {
         e->finish();
     }
-
+    deadExecutorsList.clear();
+    faabric::util::FullLock lock(mx);
```

Comment:
I'd add a newline and comment between `deadExecutorsList.clear();` and `faabric::util::FullLock lock(mx);` to make it obvious where the lock is being retaken and why. It's easy to miss this in the block as it is.